### PR TITLE
Add transformer decoder for unilm

### PR DESCRIPTION
**Pull request by FrostML**
### PR types

Others

### PR changes

Others

### Description

Add transformer decoder for unilm in Faster Transformer patches. Sub-PR of #569

### CODE REVIEW

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h [hunk 1]
```diff
@@ -0,0 +1,359 @@
...
+                    m,
+                    n);
+#ifndef NDEBUG
+      cudaDeviceSynchronize();
+      check_cuda_error(cudaGetLastError());
+#endif
+
+      if (!normalization_before_) {
+        masked_output_buf_ = norm_masked_output_buf_;
+      }
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: `normalization_before=false`的相关内容还没有完成是吧

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+    query_buf_ = buf + buf_size;  // store the query values (from_tensor * Q) in
+                                  // both masked and multi-head attention
+    key_buf_ = buf + 2 * buf_size;
+    value_buf_ = buf + 3 * buf_size;
+    context_buf_ = buf + 4 * buf_size;  // store the context result
+                                        // (softmax(qk)v) in both masked and
+                                        // multi-head attention
+
+    masked_output_buf_ = buf + 5 * buf_size;      // masked_attention_output
+    masked_output_tmp_buf_ = buf + 5 * buf_size;  // masked_attention_output
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: 这个`masked_output_tmp_buf_`是想要和`masked_output_buf_`一样吗

File: paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu [hunk 3]
```diff
@@ -0,0 +1,2672 @@
...
+    if (ite == step - 1 && tid < size_per_head) {
+      key = key_buf[qkv_id] + self_K_bias[qkv_bias_id];
+      key_cache[ite * offset + qkv_id] = key;
+    }
+
+    // T val = (tid < size_per_head) ? key * sq[tid] * (T)(scalar) : (T)(0.0f);
+    T val = (T)(0.0f);
+    if (tid < size_per_head &&
+        ite >= (start_len - memory_sequence_length[bid])) {
+      val = key * sq[tid] * (T)(scalar);
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu]: 感觉对start_len、memory_sequence_length的attention mask处理有些复杂了，是否只要下面logits计算部分那里处理就可以了

File: paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu [hunk 4]
```diff
@@ -0,0 +1,2672 @@
...
+        sum += value * logits[ite];
+      }
+    }
+    context_buf[qkv_id] = sum;
+  }
+}
+
+
+template <typename T>
+void self_attention_dispatch(const int* memory_sequence_length,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu]: 如果以这种新增函数的形式进行扩展，可否将新增的内容拎出来放到一个文件中，CMAKE的patch command中加入新增内容和编译，尽量不要带上FT中原有的内容，这样会清楚好review些

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h (hunk 1 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: 经讨论，目前支持 `normalization_before=false` 的方式正确，但是最好根据 python 代码的逻辑执行，便于 python 接口封装。

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h (hunk 2 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: 是的，用作 `normalization_before=false` 的情况下，替换回之前的指针地址。

File: paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu (hunk 3 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu (hunk 4 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu]: 单独拆开会有问题，nv 代码组织结构上的遗留问题，经讨论，采取方案，单独拎出来一个文件，cmake 进行编译的时候，将拎出来的内容重新写到 Faster Transformer 相关文件的最后进行编译。

Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/cuda/open_decoder.cu]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h (hunk 1 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+                                  step + start_len,
+                                  start_len);
+      }
+#ifndef NDEBUG
+      cudaDeviceSynchronize();
+      check_cuda_error(cudaGetLastError());
+#endif
+      decoder_norm2(from_tensor,
+                    param_.ffn_layernorm.gamma,
+                    param_.ffn_layernorm.beta,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: 这些参数也要对上，如果用post-normalization的话，这里是self attention部分layer norm的参数

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+            ffn_out_buf_,
+            m,
+            4 * n,
+            n,
+            ActivationType::GELU);
+#ifndef NDEBUG
+        cudaDeviceSynchronize();
+        check_cuda_error(cudaGetLastError());
+#endif
+        add_bias_input(ffn_out_buf_, masked_output_buf_, m, n);
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: 上面和下面这些指针交换是必要的吗，直接`add_bias_input(ffn_out_buf_, norm_masked_output_buf_, m, n);`

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h (hunk 5 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h (hunk 6 above)
Comment (FrostML) [reply paddlenlp/ops/patches/FasterTransformer/transformer_decoder.h]: 后面都是跟的 gemm 当前情况下非必要。
Done. Thanks.

File: paddlenlp/ops/patches/FasterTransformer/cuda/transformer_decoder.cu [hunk 7]
```diff
@@ -0,0 +1,643 @@
...
+                             cudaStream_t stream) {
+  const int block_sz = ATTENTION_BLOCK_SIZE;
+  T scalar = (T)(1.f / sqrtf(size_per_head * 1.0f));
+
+  dim3 grid(batch_size * head_num);
+
+  int cond = size_per_head * ((ATTENION_OPT) ? 1 : 0);
+  switch (cond) {
+    /*case 32:
+      masked_attention_kernel_opt<32, block_sz, T><<<grid, block_sz,
```
Comment (guoshengCS) [inline paddlenlp/ops/patches/FasterTransformer/cuda/transformer_decoder.cu:137]: 这里后面可能还是需要打开的

**Review by guoshengCS — APPROVED** [review]
LGTM
